package/MDAnalysis/lib/include/batched_distances.h

```diff
@@ -0,0 +1,97 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode:nil; -*- */
+/* vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4 */
+/*
+  MDAnalysis --- http://mdanalysis.googlecode.com
+
+  Copyright (c) 2006-2022 Naveen Michaud-Agrawal,
+                Elizabeth J. Denning, Oliver Beckstein,
+                and contributors (see AUTHORS for the full list)
+  Released under the GNU Public Licence, v2 or any higher version
+
+  Please cite your use of MDAnalysis in published work:
+
+      N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and
+      O. Beckstein. MDAnalysis: A Toolkit for the Analysis of
+      Molecular Dynamics Simulations. J. Comput. Chem. 32 (2011), 2319--2327,
+      in press.
+*/
+
+#ifndef __BATCHED_DISTANCES_H
+#define __BATCHED_DISTANCES_H
+
+#include <cstdint>
+#include "iterators.h"
+
+/* Functions in this header file calculate distances using batched algorithms
+ with stack allocated memory owned by the function. Testing indicated that
+ batching the calculation was much more performant than using a `.next()`
+ style API, with contiguity of memory access of paramount importance.
+ Optimal sizing of the batching is yet to be fully explored.
+
+ Functions in this header accept the iterators defined in `iterators.h`
+ (currently _AtomGroupIterator and _ArrayIterator) that provide a homogenous
+ interface for iterating through their respective datatypes
+ (AtomGroups and NumPy ndarrays).
+
+ Templates are used to allow iterators at either position, avoiding writing
+ multiple overloads for functions that accept many iterators.
+
+*/
+
+template <typename T, typename U>
+void _calc_distance_array_batched(T ref, U conf, double *distances, uint64_t batchsize)
+{
+    // batchsize is in particles, so allocate buffers of 3x batchsize.
+    const uint64_t atom_bufsize = 3 * batchsize;
+    float ref_buffer[atom_bufsize];
+    float conf_buffer[atom_bufsize];
+
+    // slight indirection required here to homogenize interface between
+    // _AtomGroupIterator and _ArrayIterator where passing stack allocated
+    // array as float*& does not decay to a float* as desired.
+    float *ref_buffer_ = ref_buffer;
+    float *conf_buffer_ = conf_buffer;
```

Comment:
Could the scope of these be reduced to inside the loop if `load_into` returned a float*?

Reply:
I will investigate

Reply:
Done, I think not passing the pointer by reference is advisable in general as risk for modification is high.